cpp/src/io/parquet/rle_stream.cuh

```diff
@@ -0,0 +1,353 @@
+      int const lead_values      = (run_offset - effective_offset);
+      output_pos -= lead_values;
+      remain += lead_values;
+      cur = run_start + ((effective_offset >> 3) * level_bits);
+    }
+
+    // if this is a repeated run, compute the repeated value
+    int _level_val;
+    if (!(level_run & 1)) {
+      _level_val = run_start[0];
+      if (level_bits > 8) { _level_val |= run_start[0] << 8; }
+    }
+
+    // process
+    while (remain > 0) {
+      int batch_len = min(32, remain);
+
+      // if this is a literal run. each thread computes it's own level_val
+      if (level_run & 1) {
+        int const batch_len8 = (batch_len + 7) >> 3;
+        if (lane < batch_len) {
+          int bitpos                = lane * level_bits;
+          uint8_t const* cur_thread = cur + (bitpos >> 3);
+          bitpos &= 7;
+          _level_val = 0;
+          if (cur_thread < end) { _level_val = cur_thread[0]; }
+          cur_thread++;
+          if (level_bits > 8 - bitpos && cur_thread < end) {
+            _level_val |= cur_thread[0] << 8;
+            cur_thread++;
+            if (level_bits > 16 - bitpos && cur_thread < end) { _level_val |= cur_thread[0] << 16; }
+          }
+          _level_val = (_level_val >> bitpos) & ((1 << level_bits) - 1);
+        }
+
+        cur += batch_len8 * level_bits;
+      }
+
+      // store level_val
+      if (lane < batch_len && (lane + output_pos) >= 0) { output[lane + output_pos] = _level_val; }
+      remain -= batch_len;
+      output_pos += batch_len;
+    }
+  }
+};
+
+// a single rle run. may be broken up into multiple rle_batches
+template <typename level_t>
+struct rle_run {
+  int size;  // total size of the run
+  int output_pos;
+  uint8_t const* start;
+  int level_run;  // level_run header value
+  int remaining;
+
+  __device__ __inline__ rle_batch<level_t> next_batch(level_t* const output, int max_size)
+  {
+    int batch_len        = min(max_size, remaining);
+    int const run_offset = size - remaining;
+    remaining -= batch_len;
+    return rle_batch<level_t>{start, run_offset, output, level_run, batch_len};
+  }
+};
+
+// a stream of rle_runs
+template <typename level_t>
+struct rle_stream {
+  int level_bits;
+  uint8_t const* start;
+  uint8_t const* cur;
+  uint8_t const* end;
+
+  int max_output_values;
+  int total_values;
+  int cur_values;
+
+  level_t* output;
+
+  rle_run<level_t>* runs;
+  int run_index;
+  int run_count;
+  int output_pos;
+  bool spill;
+
+  int next_batch_run_start;
+  int next_batch_run_count;
+
+  __device__ rle_stream(rle_run<level_t>* _runs) : runs(_runs) {}
+
+  __device__ void init(int _level_bits,
+                       uint8_t const* _start,
+                       uint8_t const* _end,
+                       int _max_output_values,
+                       level_t* _output,
+                       int _total_values)
+  {
+    level_bits = _level_bits;
+    start      = _start;
+    cur        = _start;
+    end        = _end;
```

Comment:
don't we usually add the prefix to data members instead?

Reply:
In my experience, it's the reverse.   Otherwise you have _ propagating throughout tons of code instead of just at initialization.

Reply:
Plus here it's a struct...kinda weird to have prefixes on public members.

Reply:
Ah, that's right.